flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoderDelegateTest.java

```diff
@@ -0,0 +1,324 @@
+		}
+
+		// Partitions the remaining data into three parts averagely.
+		int remainingBytes = buffer.readableBytes() - position;
+
+		position += (remainingBytes / 3);
+		partitionPositions.add(position);
+
+		position += (remainingBytes / 3 * 2);
+		partitionPositions.add(position);
+
+		partitionPositions.add(buffer.readableBytes());
+
+		// Partitions the buffers according to the selected partition positions.
+		List<ByteBuf> result = new ArrayList<>();
+		for (int i = 1; i < partitionPositions.size(); ++i) {
+			ByteBuf partitionedBuffer = ALLOCATOR.buffer();
+			partitionedBuffer.writeBytes(
+				buffer,
+				partitionPositions.get(i - 1),
+				partitionPositions.get(i) - partitionPositions.get(i - 1));
+			result.add(partitionedBuffer);
+		}
+
+		return result;
+	}
+
+	private List<NettyMessage> decodedMessages(EmbeddedChannel channel, List<ByteBuf> inputBuffers) {
+		for (ByteBuf buffer : inputBuffers) {
+			channel.writeInbound(buffer);
+		}
+
+		channel.runPendingTasks();
+
+		List<NettyMessage> decodedMessages = new ArrayList<>();
+		Object input;
+		while ((input = channel.readInbound()) != null) {
+			assertTrue(input instanceof NettyMessage);
+			decodedMessages.add((NettyMessage) input);
+		}
+
+		return decodedMessages;
+	}
+
+	private void verifyDecodedMessages(List<NettyMessage> expected, List<NettyMessage> decodedMessages) {
+		assertEquals(expected.size(), decodedMessages.size());
+		for (int i = 0; i < expected.size(); ++i) {
+			assertEquals(expected.get(i).getClass(), decodedMessages.get(i).getClass());
+
+			if (expected.get(i) instanceof NettyMessage.BufferResponse) {
+				BufferResponse expectedBufferResponse = (BufferResponse) expected.get(i);
+				BufferResponse decodedBufferResponse = (BufferResponse) decodedMessages.get(i);
+
+				assertEquals(expectedBufferResponse.backlog, decodedBufferResponse.backlog);
+				assertEquals(expectedBufferResponse.sequenceNumber, decodedBufferResponse.sequenceNumber);
+				assertEquals(expectedBufferResponse.bufferSize, decodedBufferResponse.bufferSize);
+				assertEquals(expectedBufferResponse.receiverId, decodedBufferResponse.receiverId);
+
+				if (expectedBufferResponse.bufferSize == 0 ||
+					expectedBufferResponse.receiverId.equals(RELEASED_CHANNEL_ID) ||
+					expectedBufferResponse.receiverId.equals(REMOVED_CHANNEL_ID)) {
+
+					assertNull(decodedBufferResponse.getBuffer());
+				} else if (expectedBufferResponse.receiverId.equals(NORMAL_CHANNEL_ID)) {
+					assertEquals(expectedBufferResponse.getBuffer(), decodedBufferResponse.getBuffer());
+				} else {
+					fail("The decoded buffer response was sent to unknown input channel.");
+				}
+			} else if (expected.get(i) instanceof NettyMessage.ErrorResponse) {
+				ErrorResponse expectedErrorResponse = (ErrorResponse) expected.get(i);
+				ErrorResponse decodedErrorResponse = (ErrorResponse) decodedMessages.get(i);
+
+				assertEquals(expectedErrorResponse.receiverId, decodedErrorResponse.receiverId);
```

Comment:
ditto: also for the case of `ErrorResponse` to extract a static method

Reply:
Extractor the comparison of error response.